src/webots/nodes/WbCadShape.cpp

```diff
@@ -575,45 +575,49 @@ const WbVector3 WbCadShape::absoluteScale() const {
   WbField urlFieldCopy(*findField("url", true));
   for (int i = 0; i < mUrl->size(); ++i) {
-    if (WbUrl::isLocalUrl(mUrl->value()[i])) {
-      dynamic_cast<WbMFString *>(urlFieldCopy.value())->setItem(i, WbUrl::computeLocalAssetUrl(this, "url", mUrl->value()[i]));
-    } else if (WbUrl::isWeb(mUrl->value()[i]))
-      continue;
+    const QString &completeUrl = WbUrl::computePath(this, "url", mUrl, i);
+    if (WbUrl::isLocalUrl(completeUrl))
+      dynamic_cast<WbMFString *>(urlFieldCopy.value())->setItem(i, WbUrl::computeLocalAssetUrl(completeUrl));
+    else if (WbUrl::isWeb(completeUrl))
+      dynamic_cast<WbMFString *>(urlFieldCopy.value())->setItem(i, completeUrl);
     else {
-      const QString meshPath(WbUrl::computePath(this, "url", mUrl, i));
-      if (writer.isWritingToFile()) {
-        QString newUrl = WbUrl::exportMesh(this, mUrl, i, writer);
-        dynamic_cast<WbMFString *>(urlFieldCopy.value())->setItem(i, newUrl);
-      }
-
-      const QString &url(mUrl->item(i));
-      writer.addResourceToList(url, meshPath);
+      QString meshPath;
+      if (writer.isWritingToFile())
+        dynamic_cast<WbMFString *>(urlFieldCopy.value())->setItem(i, WbUrl::exportMesh(this, mUrl, i, writer));
+      else
+        dynamic_cast<WbMFString *>(urlFieldCopy.value())->setItem(i, WbUrl::expressRelativeToWorld(completeUrl));
     }
   }
-  const QString &completeUrl = WbUrl::computePath(this, "url", mUrl->item(0));
-  const QString prefix = completeUrl.left(completeUrl.lastIndexOf('/'));
-  for (QString material : objMaterialList(completeUrl)) {
-    QString newUrl;
-    if (writer.isWritingToFile())
-      newUrl = WbUrl::exportResource(this, material, WbUrl::computePath(this, "url", mUrl, 0), writer.relativeMeshesPath(),
-                                     writer, false);
-    else
-      newUrl = prefix + '/' + material;
-
-    dynamic_cast<WbMFString *>(urlFieldCopy.value())->addItem(newUrl);
-    writer.addResourceToList(newUrl, newUrl);
+
+  // export materials
+  if (writer.isX3d()) {  // only needs to be included in the x3d, when converting to base node it shouldn't be included
+    const QString &parentUrl = WbUrl::computePath(this, "url", mUrl->item(0));
+    for (QString material : objMaterialList(parentUrl)) {
+      QString materialUrl = WbUrl::combinePaths(material, parentUrl);
+      if (WbUrl::isLocalUrl(materialUrl))
+        dynamic_cast<WbMFString *>(urlFieldCopy.value())->addItem(WbUrl::computeLocalAssetUrl(materialUrl));
+      else if (WbUrl::isWeb(materialUrl))
+        dynamic_cast<WbMFString *>(urlFieldCopy.value())->addItem(materialUrl);
+      else {
+        if (writer.isWritingToFile())
+          dynamic_cast<WbMFString *>(urlFieldCopy.value())
+            ->addItem(WbUrl::exportResource(this, materialUrl, materialUrl, writer.relativeMeshesPath(), writer, false));
+        else
+          dynamic_cast<WbMFString *>(urlFieldCopy.value())->addItem(WbUrl::expressRelativeToWorld(materialUrl));
+      }
```

Comment:
More readable and less error prone:
```suggestion
      WbMFString *urlField = dynamic_cast<WbMFString *>(urlFieldCopy.value());
      if (WbUrl::isLocalUrl(materialUrl))
        urlField->addItem(WbUrl::computeLocalAssetUrl(materialUrl));
      else if (WbUrl::isWeb(materialUrl))
        urlField->addItem(materialUrl);
      else {
        if (writer.isWritingToFile())
          urlField->addItem(WbUrl::exportResource(this, materialUrl, materialUrl, writer.relativeMeshesPath(), writer, false));
        else
          urlField->addItem(WbUrl::expressRelativeToWorld(materialUrl));
      }
```